src/Stratis.Bitcoin.Features.Miner/PosMinting.cs

```diff
@@ -363,7 +373,8 @@ public class CoinstakeContext
         public IAsyncLoop Stake(WalletSecret walletSecret)
```

Comment:
Personally, I am not a big fan of "hiding" logic this. I prefer to code with intent. I.e. if we call the Stake method, we stake. We shouldnt check inside the Stake method if staking is running. Preferably when we call Stake() and we are staking, the method should throw and exception.

I know we have it like this because of the Miner controller but:

`public IActionResult StartStaking([FromBody]StartStakingRequest request)`

should check to see if we are staking before call Stake(), e.g. 

```
var staking = this.fullNode.NodeFeature<MiningFeature>(true);
if(staking.CanStake)
    staking.StartStaking(request.Name, request.Password);
```

This is more intuitive and removes the need for thread safety checks etc.

Reply:
I agree. Having `StartStaking`\`StartStakingContinously` in the miner sounds like a more logical method name. 

We probably can rename controller's `StartStaking` to smth like RequestStartStaking as well.

Reply:
>should check to see if we are staking before call Stake(), e.g.

>if(staking.CanStake)
    staking.StartStaking(request.Name, request.Password);

>This is more intuitive and removes the need for thread safety checks etc.

But the ` public IActionResult StartStaking([FromBody]StartStakingRequest request)` can be called twice at the same time so there is a race condition and we still need the thread safety checks.

Reply:
Yes, so that is what a possible `CanStake` property will be used for. The moment that the asyncloop starts up in the Stake() method, we set CanStake to false. I agree, we still need to make the start of the loop thread safe.

The point that I'm also trying to make is that we should expose whether or not we can can stake so that it can interrogated by whatever consumer. It shouldn't be hidden. Therefore, we should first check IF we can stake and then call Stake(). The Stake() method shouldn't be responsible for determining whether it should stake or not ;)

Reply:
>Yes, so that is what a possible CanStake property will be used for. The moment that the asyncloop starts up in the Stake() method, we set CanStake to false. I agree, we still need to make the start of the loop thread safe though.

To make it thread safe we'd need to call CanStake check and Stake() inside a lock. 
So I guess  I'm fine with the safety checks inside the Stake() since this way we won't make the caller take care of thread safety. 

>The point that I'm also trying to make is that we should expose whether or not we can can stake so that it can interrogated by whatever consumer.

Yeah, that would be nice to have.

Reply:
> To make it thread safe we'd need to call CanStake check and Stake() inside a lock.

Hmm you do have a point here! Just looking at the code again... I see that when the node starts up, we start the MiningFeature and subsequenty PosMinting IF the stake arg was passed. So in theory we can only start and stop staking if it was enabled on node start up. Also, seeing as this a wallet call, I can't see how multiple threads will be able to call this method?

But I suppose to make it air tight we need to ensure that it can't happen. Ever!

Reply:
> So in theory we can only start and stop staking if it was enabled on node start up. 

Not really. Enabling it by using "-stake" will just make node call `MiningFeature.StartStaking` on startup. 

But when we call StartStaking from the controller there is no check for the `-stake` startup option to be == true.

>Also, seeing as this a wallet call, I can't see how multiple threads will be able to call this method?

Imagine that someone uses the wallet and presses the start staking button 2 times- 2 calls of MinerController.StartStaking. Here is the race condition so it's possible that both of the calls will pass the .CanStake check and both of them call StartStaking(). 
That's why I think it's actually a good idea to have thread safety checks inside the StartStaking() somewhere like @dangershony  did this.

Reply:
@fassadlr there is still a possibility of a race if we don't lock the entry point to the method, this method is called form the controller and so is a high candidate for synchronisation issues.